src/fastcs/backend.py

```diff
@@ -0,0 +1,140 @@
+import asyncio
+from collections import defaultdict
+from typing import Callable, cast
+
+from .attributes import AttrCallback, AttrMode, AttrR, AttrW
+from .cs_methods import MethodType
+from .mapping import Mapping, SingleMapping
+
+
+def _get_initial_tasks(mapping: Mapping) -> list[Callable]:
+    initial_tasks: list[Callable] = []
+    initial_tasks.append(mapping.controller.connect)
+    return initial_tasks
+
+
+def _create_periodic_scan_task(period, methods: list[Callable]) -> Callable:
+    async def scan_task() -> None:
+        while True:
+            await asyncio.gather(*[method() for method in methods])
+            await asyncio.sleep(period)
+
+    return scan_task
+
+
+def _get_periodic_scan_tasks(scan_dict: dict[float, list[Callable]]) -> list[Callable]:
+    periodic_scan_tasks: list[Callable] = []
+    for period, methods in scan_dict.items():
+        periodic_scan_tasks.append(_create_periodic_scan_task(period, methods))
+
+    return periodic_scan_tasks
+
+
+def _add_wrapped_scan_tasks(
+    scan_dict: dict[float, list[Callable]], single_mapping: SingleMapping
+):
+    for method_data in single_mapping.methods:
+        if method_data.info.method_type == MethodType.scan:
+            period = method_data.info.kwargs["period"]
+            method = method_data.method
+            scan_dict[period].append(method)
+
+
+def _create_updater_callback(attribute, controller):
+    async def callback():
+        await attribute.updater.update(controller, attribute)
+
+    return callback
+
+
+def _add_updater_scan_tasks(
+    scan_dict: dict[float, list[Callable]], single_mapping: SingleMapping
+):
+    for attribute in single_mapping.attributes.values():
+        if attribute.mode in (AttrMode.READ, AttrMode.READ_WRITE):
+            attribute = cast(AttrR, attribute)
+
+            if attribute.updater is None:
+                continue
+
+            callback = _create_updater_callback(attribute, single_mapping.controller)
+            scan_dict[attribute.updater.update_period].append(callback)
+
+
+def _get_scan_tasks(mapping: Mapping) -> list[Callable]:
+    scan_dict: dict[float, list[Callable]] = defaultdict(list)
+
+    for single_mapping in mapping.get_controller_mappings():
+        _add_wrapped_scan_tasks(scan_dict, single_mapping)
+        _add_updater_scan_tasks(scan_dict, single_mapping)
+
+    scan_tasks = _get_periodic_scan_tasks(scan_dict)
+    return scan_tasks
+
+
+def _link_single_controller_put_tasks(single_mapping: SingleMapping):
+    put_methods = [
+        method_data
+        for method_data in single_mapping.methods
+        if method_data.info.method_type == MethodType.put
+    ]
+
+    for method_data in put_methods:
+        method = cast(AttrCallback, method_data.method)
+        name = method_data.name.removeprefix("put_")
+
+        attribute = single_mapping.attributes[name]
```

Comment:
Is it automatic that there is an attribute corresponding to `put_*` or does the user have to get this right and it will crash otherwise?

Reply:
It will raise a KeyError, looking at the code.

I am concerned about the use of `@put` wrappers; I wasn't just disagreeing with Tom. I'm just not sure what it would look like to rely _only_ on Sender classes. Perhaps it would be best to force our testers to use Sender classes for the moment, and see when they get stuck. Easy to remove this code while `@scan` and `@command` can remain as wrappers.

What do you think?

Reply:
As discussed. I think probably we don't need `@put`.

Reply:
We actually decided that we would keep both for now!